### DISCUSSION TITLE
Does count of workflows stuck in delayed queue , impact performance of ST2 execution?

### DISCUSSION THREAD

**Discussion by zsmanjot** [StackStorm/st2#6247] [Operations]
I have an environment where we have tons of alerts and those are actioned upon by ST2. But when alert flow is more and we have delayed policy in place , the workflows get stuck in delayed queue and the execution of other workflows becomes very slow.

For example:
<img width="871" alt="image" src="https://github.com/user-attachments/assets/04d4d1a0-a97a-4b6d-bf22-d33dfcf6ac3d">

And here below one execution started at 19 sept 16:38 UTC but completed at 20 Sept 00:15 UTC

Almost after 8 hours, the execution got completed.

<img width="866" alt="image" src="https://github.com/user-attachments/assets/9c07140a-b052-4838-8469-b9fbc231e627">
 
<img width="910" alt="image" src="https://github.com/user-attachments/assets/19b8fbe3-6cc9-4ec0-a8d6-564f3ff72949">

Anyone who can suggest here?

**Comment by orangelzh** [comment]
how many action runner do you have
and how many workflowengine process do you run
suggest imcrease this two process number。

**Comment by zsmanjot** [comment]
Have 20 action runners, How can we check workflow engine process?

**Reply by orangelzh** [reply]
ps auxf|grep workflowengine
The recommended ratio of workflowengine to actionrunner is 1:3
If the number of action tasks to be processed concurrently is 100, it is recommended that the number of action runners is 50 (assuming that it takes 30 seconds to execute an action), and the number of workflowengines is 20.

the following script is start workflowengine process：
#!/bin/bash
while [[ $retry_i -lt 20 ]]
do
now=`date +%Y%m%d%H%M`
retry_i=`expr $retry_i + 1 `
if [[ `ps auxf|grep st2workflowengine |wc -l ` -gt 20 ]]
then
  exit
fi
nohup /opt/stackstorm/st2/bin/python /opt/stackstorm/st2/bin/st2workflowengine --config-file /etc/st2/st2.conf >> /var/log/st2/st2workflowengine_${now}_${retry_i}.log  2>&1 &
sleep 2
done

**Comment by zsmanjot** [comment]
Hey i have 200 actionrunners but just 2 workflowengine running. How can i increase the st2workflowengine? Please help.

**Reply by orangelzh** [reply]
#!/bin/bash
retry_i=1
while [[ $retry_i -lt 60 ]]
do
now=date +%Y%m%d%H%M
retry_i=expr $retry_i + 1 
if [[ ps auxf|grep st2workflowengine |wc -l  -gt 60 ]]
then
exit
fi
nohup /opt/stackstorm/st2/bin/python /opt/stackstorm/st2/bin/st2workflowengine --config-file /etc/st2/st2.conf >> /var/log/st2/st2workflowengine_${now}_${retry_i}.log 2>&1 &
sleep 2
done

**Comment by zsmanjot** [comment]
@orangelzh 

Could you explain how it is going to help? Where i have to place this script and how it will be invoked?

**Reply by orangelzh** [reply]
Just execute this script directly on your st2 machine, and then ps auxf | grep workflowengine to see if the number of processes is correct.
